base/applications/rapps/appinfo.cpp

```diff
@@ -561,15 +574,54 @@ CInstalledApplicationInfo::GetDisplayInfo(CStringW &License, CStringW &Size, CSt
     ATLASSERT(FALSE && "Should not be called");
 }
 
+InstallerType
+CInstalledApplicationInfo::GetInstallerType() const
+{
+    CRegKey reg;
+    if (reg.Open(m_hKey, GENERATE_ARPSUBKEY, KEY_READ) == ERROR_SUCCESS)
+    {
+        return INSTALLER_GENERATE;
+    }
+    return INSTALLER_UNKNOWN;
+}
+
 BOOL
-CInstalledApplicationInfo::UninstallApplication(BOOL bModify)
+CInstalledApplicationInfo::UninstallApplication(UINT Flags)
 {
+    extern BOOL
+    UninstallGenerated(CInstalledApplicationInfo &AppInfo, UINT Flags);
+
+    if (GetInstallerType() == INSTALLER_GENERATE)
+    {
+        return UninstallGenerated(*this, Flags);
+    }
+
+    BOOL bModify = Flags & UCF_MODIFY;
     if (m_szUninstallString.IsEmpty())
     {
         RetrieveUninstallStrings();
     }
 
-    BOOL bSuccess = StartProcess(bModify ? m_szModifyString : m_szUninstallString, TRUE);
+    CStringW cmd = bModify ? m_szModifyString : m_szUninstallString;
+    if ((Flags & (UCF_MODIFY | UCF_SILENT)) == UCF_SILENT)
+    {
+        DWORD msi = 0;
```

Comment:
What's msi?

Reply:
The next line is `msi = GetApplicationRegDword(L"WindowsInstaller", &msi) && msi;`, it should be pretty clear what MSI means.